### ISSUE TITLE
NPE in TopicFragmentFilter module [DITA OT 2.x]

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2222]
DITA Map:

``` xml
<!DOCTYPE map PUBLIC "-//OASIS//DTD DITA Map//EN" "map.dtd">
<map>
 <title>DITA Topic Map</title>
 <keydef href="zuzu.dita" keys="zuzu"/>
 <topicref href="test.dita"/>
</map>
```

zuzu.dita:

``` xml
<!DOCTYPE dita PUBLIC "-//OASIS//DTD DITA Composite//EN" "ditabase.dtd">
<dita>
  <topic id="a">
    <title>A</title>
  </topic>
  <topic id="Z">
    <title>A</title>
  </topic>
</dita>
```

test.dita:

``` xml
<!DOCTYPE topic PUBLIC "-//OASIS//DTD DITA Topic//EN" "topic.dtd">
<topic id="topic_fyk_zzc_x5">
  <title>AAA</title>
  <body/>
  <topic id="id_kq2_11d_x5" conkeyref="zuzu" conrefend="default.dita#default/Z">
    <title/>
  </topic>
</topic>
```

NPE when publishing:

```
BUILD FAILED
D:\projects\eXml\frameworks\dita\DITA-OT2.x\build.xml:41: The following error occurred while executing this line:
D:\projects\eXml\frameworks\dita\DITA-OT2.x\plugins\org.dita.base\build_preprocess.xml:226: java.lang.NullPointerException
    at java.util.ArrayDeque.addFirst(Unknown Source)
    at org.dita.dost.writer.TopicFragmentFilter.startElement(TopicFragmentFilter.java:48)
    at org.apache.xerces.parsers.AbstractSAXParser.startElement(Unknown Source)
    at org.ditang.relaxng.defaults.RelaxNGDefaultsComponent.startElement(RelaxNGDefaultsComponent.java:200)
    at org.apache.xerces.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:268)
    at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(XMLDocumentFragmentScannerImpl.java:1655)
    at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:325)
```

**Comment by robander** [comment]
Marking as a bug because any NPE is a bug - but in this case the NPE is caused by invalid markup. The conref end value is set to `conrefend="default.dita#default/Z"`, which means "The element with `id="Z"` inside the topic with `id="default"`, and that does not exist.

**Comment by raducoravu** [comment]
Actually the specs:

    https://docs.oasis-open.org/dita/v1.2/os/spec/common/theconrefendattribute.html

says:

    When the conkeyref attribute is used in place of conref, a key is used to address the target of the reference. The conrefend attribute, which indicates the end of a conref range, may not use a key. Instead the the map or topic element addressed by the key name component of the conkeyref is used in place of whatever map or topic element is addressed by the conrefend attribute.

So when using conkeyref, the conrefend attribute just needs to have the proper pattern, the only bit from it used is the element ID. So my example is valid.

**Comment by robander** [comment]
That ... is odd. I'm sure that was written in the context of elements-within-topics, without regard to topic elements. The value `"default.dita#default/Z"` on its own would not be valid, for the reason given above. If working with something like `<section>` or `<li>` elements, the quoted text from the spec would make sense, replacing `default.dita#default` with the topic at the start of the range. But in this case, you're referencing actual topics, so you can't actually use "the map or topic element addressed by the key name", because you end up with the`topicID/topicID` construct.

This all just feels ... odd. I feel sure this wasn't considered when working on that language about ranges.

**Comment by raducoravu** [comment]
Looked again at my example and now it also looks wrong to me. Maybe the bottom line would be the fact that you cannot use conkeyref range to pull a range of topic elements. And the NPE would need to be guarded somehow.

**Comment by jelovirt** [comment]
@robander should we throw an error for `conrefend="default.dita#default/Z"` because `@conrefend` points to another topic? Just using `Z`, even with a warning, feels too aggressive error recovery.

**Comment by robander** [comment]
I think so. Probably should raise this at the TC - I think this is probably a case that was never considered, and I'm not confident of a quick working solution that would allow it.

**Comment by robander** [comment]
Still get an NPE from this in 3.1.
[2222.zip](https://github.com/dita-ot/dita-ot/files/2246222/2222.zip)

**Comment by robander** [comment]
Ahh - this isn't actually related to conkeyref, or even (directly) to the topic fragment module.

When we evaluate a conref range, the code processes the first element correctly, and then processes everything in sequence from there up to the end. For each node in that sequence (including any nodes in the middle + the end node), the `@id` attribute is explicitly ignored -- which is correct for most elements. But in this case, it means we generate `<topic>` elements without an ID. Many DITA-OT modules assume every topic has an ID (it's a pretty basic requirement). The topic fragment module is just the first post-conref one we get to that explicitly tries to read `topic/@id` and doesn't have exception handling in place for a missing ID.

**Comment by robander** [comment]
Fixed with #3033
